File: Day-01-Arrays/JavaScript/121_Best_Time_to_Buy_and_Sell_Stock.js
/*
Problem: 121. Best Time to Buy and Sell Stock

Pattern: Running Minimum / Greedy

Time Complexity: O(n)
Space Complexity: O(1)
*/

var maxProfit = function(prices) {

    let minPrice = prices[0];
    let maxProfit = 0;

    for (let i = 1; i < prices.length; i++) {

        minPrice = Math.min(minPrice, prices[i]);

        const profit = prices[i] - minPrice;

        maxProfit = Math.max(maxProfit, profit);
    }

    return maxProfit;
};

File: Day-01-Arrays/Python/121_Best_Time_to_Buy_and_Sell_Stock.py
"""
Problem: 121. Best Time to Buy and Sell Stock

Pattern: Running Minimum / Greedy

Time Complexity: O(n)
Space Complexity: O(1)
"""

class Solution:
    def maxProfit(self, prices: list[int]) -> int:

        minPrice = prices[0]
        maxProfit = 0

        for price in prices[1:]:

            minPrice = min(minPrice, price)
            profit = price - minPrice
            maxProfit = max(maxProfit, profit)

        return maxProfit